### PR TITLE
feat: Add glob patterns support

### PR DESCRIPTION
**Pull request by JinnyYi**
ref: https://github.com/beyondstorage/beyond-ctl/issues/98

### CODE REVIEW

File: operations/ls.go [hunk 1]
```diff
@@ -2,37 +2,166 @@ package operations
...
+		} else {
+			objects = append(objects, &types.Object{
+				Mode: types.ModeDir,
+				Path: workPath,
+				ID:   workPath,
+			})
+		}
+
+		return objects, nil
+	}
```
Comment (JinnyYi) [inline operations/ls.go:127]: I'm confused about whether the `GlobList` (to expand glob pattern) should return `objects []*types.Object` or `files []string`. If returns `[]string`, we need to repeat the `list`. If returns `[]*types.Object`, the above processing looks strange.

Comment (Xuanwo) [reply operations/ls.go:127]: It does look strange.

How about keeping `List` unchanged, and implementing a `ListWithGlob(path, pattern string)`?

In `ListWith`

Comment (Xuanwo) [reply operations/ls.go:127]: Why we need a `DirWalk` here?

File: go.mod [hunk 2]
```diff
@@ -6,6 +6,7 @@ require (
 	github.com/BurntSushi/toml v0.4.1
 	github.com/Xuanwo/go-bufferpool v0.2.0
 	github.com/docker/go-units v0.4.0
+	github.com/gobwas/glob v0.2.4-0.20181002190808-e7a84e9525fe
```
Comment (Xuanwo) [inline go.mod:9]: Please always use tagged version unless we don't have choice.

File: operations/ls.go (hunk 1 above)
Comment (JinnyYi) [reply operations/ls.go:127]: I'm trying to recursively expand the pattern in `DirWalk`.

File: go.mod (hunk 2 above)
Comment (JinnyYi) [reply go.mod:9]: Got it.

File: operations/ls.go [hunk 3]
```diff
@@ -75,31 +76,27 @@ func (so *SingleOperator) listRecursively(
...
-		c := path[i]
-		if c == '\\' {
-			i++
-		} else if c == '/' {
-			splitIdx = i
-		} else if c == '*' || c == '?' || c == '[' || c == '{' {
-			hasMeta = true
-			break
-		}
+func hasMeta(path string) bool {
```
Comment (Xuanwo) [inline operations/ls.go:79]: Maybe we can remove this function? if `doublestar.SplitPattern(path) == path`, the path must not have meta?

File: operations/ls.go [hunk 4]
```diff
@@ -71,3 +75,158 @@ func (so *SingleOperator) listRecursively(
...
+	base, pattern := splitPattern(path)
+	it, err := so.store.List(base, pairs.WithListMode(types.ListModeDir))
+	if err != nil {
+		if errFlag {
+			ch <- &ObjectResult{Error: err}
+		}
+		return
+	}
+
+	subPatterns := strings.SplitN(pattern, "/", 2)
```
Comment (Xuanwo) [inline operations/ls.go:128]: Why do we need to split the pattern here?

File: operations/ls.go [hunk 5]
```diff
@@ -71,3 +75,158 @@ func (so *SingleOperator) listRecursively(
...
+			break
+		}
+		if err != nil {
+			if errFlag {
+				ch <- &ObjectResult{Error: err}
+			}
+			break
+		}
+
+		switch len(subPatterns) {
```
Comment (Xuanwo) [inline operations/ls.go:141]: What's the different from `len(subPatterns)`?

File: operations/ls.go [hunk 6]
```diff
@@ -71,3 +75,158 @@ func (so *SingleOperator) listRecursively(
...
+				if !o.Mode.IsDir() {
+					ch <- &ObjectResult{Object: o}
+				} else {
+					so.listWithGlob(ch, o.Path, false)
+				}
+			}
+			break
+		case 2:
+			ok1, _ := doublestar.Match(base+subPatterns[0], o.Path)
+			ok2, _ := doublestar.Match(base+subPatterns[0]+"/", o.Path)
```
Comment (Xuanwo) [inline operations/ls.go:155]: Why do we need to match twice here?

**Comment by Xuanwo** [comment]
The current implementation is still too complex for me. Can we make the implementation simpler?

File: operations/ls.go (hunk 3 above)
Comment (JinnyYi) [reply operations/ls.go:79]: In the process of expanding the path step by step, it may not contain meta at a certain level, like `test/a*/b/c*`?

**Comment by JinnyYi** [comment]
We will re-implement it as https://github.com/beyondstorage/beyond-ctl/issues/98#issuecomment-966114027 described.

**Pull request closed without merging**